KAFKA-7617: Add authorization primitives to security page

viktorsomogyi opened the pull request:
This is a security page improvement that adds documentation about Kafka authorization primitives to the security page.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

viktorsomogyi committed:
KAFKA-7617: Add protocol privileges to security page

viktorsomogyi commented:
@asasvari would you please review this?

viktorsomogyi commented:
@TolerableCoder would you review this too please? (And please point out grammatical errors as well :) )

stanislavkozlovski commented on docs/security.html:
```diff
@@ -1205,6 +1205,447 @@ <h4><a id="security_authz_cli" href="#security_authz_cli">Command Line Interface
 
+    <h4><a id="security_authz_primitives" href="#security_authz_primitives">Privilege primitives</a></h4>
+    <p>Protocol calls are usually performing some operations on certain resources in Kafka. It is required to know the
+    operations and resources to set up an effective protection. In this section we'll list these operations and
```
I'm not entirely certain on the correct grammar here, but `set up effective protection` sounds better to me

stanislavkozlovski commented on docs/security.html:
```diff
@@ -1205,6 +1205,447 @@ <h4><a id="security_authz_cli" href="#security_authz_cli">Command Line Interface
+        <li>All</li>
+    </ul>
+    <h5>Resources in Kafka</h5>
+    <p>Operations above can be applied on certain resources which are described below.</p>
```
nit: **The** operations

stanislavkozlovski commented on docs/security.html:
```diff
@@ -1205,6 +1205,447 @@ <h4><a id="security_authz_cli" href="#security_authz_cli">Command Line Interface
+        writing them) require the corresponding privilege to be added. If there is an authorization error with a
+        topic resource, then a TOPIC_AUTHORIZATION_FAILED (error code: 29) will be returned.</li>
+        <li><b>Group:</b> this represents the consumer groups in the brokers. All protocol calls that are working with
+        consumer groups, like joining to a group must have privileges with the group in subject. If the privilege is not
```
"joining to a group" -> "joining a group"

stanislavkozlovski commented on docs/security.html:
```diff
@@ -1205,6 +1205,447 @@ <h4><a id="security_authz_cli" href="#security_authz_cli">Command Line Interface
+        <li><b>Cluster:</b> this resource represents the cluster. Operations that are affecting the whole cluster, like
+        controlled shutdown are protected by privileges on the Cluster resource. If there is an authorication problem
+        on a cluster resource, then a CLUSTER_AUTHORIZATION_FAILED (error code: 31) will be returned.</li>
+        <li><b>TransactionalId:</b> this resource represents the resource which represents the transactions. Actions,
```
Would the same meaning be preserved if we say "this resource represents actions related to transactions, such as committing", instead of the first two sentences?

viktorsomogyi replied:
yea I think so. reading my phrasing again seems a bit repetitive, I'll use yours.

stanislavkozlovski commented on docs/security.html:
```diff
@@ -1205,6 +1205,447 @@ <h4><a id="security_authz_cli" href="#security_authz_cli">Command Line Interface
+                <td>FETCH (1)</td>
+                <td>ClusterAction</td>
+                <td>Cluster</td>
+                <td>A follower must have ClusterAction on Cluster resource in order to fetch partition data.</td>
```
on **the** Cluster resource

stanislavkozlovski reviewed (commented):
These changes look very good @viktorsomogyi. Thanks a lot for improving our documentation - it is always needed and very helpful.
I left a few small comments

viktorsomogyi committed:
Resolving grammatical problems and moving the section after simple acl

viktorsomogyi commented:
@stanislavkozlovski thank you very much for the quick review. I've fixed those issues and moved the section after the simple acl authorizer as it was sitting right between the simple acl sections which I didn't realize earlier.

mrsrinivas reviewed (approved):
Thanks for the PR @viktorsomogyi 

LGTM

omkreddy commented on docs/security.html:
```diff
@@ -1258,6 +1258,445 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <td>DESCRIBE_CONFIGS (32)</td>
+            <td>DescribeConfigs</td>
+            <td>Cluster</td>
+            <td>If broker configs are requested, then the broker will check topic level privileges.</td>
```
broker will check cluster level privileges?

omkreddy commented on docs/security.html:
```diff
@@ -1258,6 +1258,445 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <td>ALTER_CONFIGS (33)</td>
+            <td>AlterConfigs</td>
+            <td>Cluster</td>
+            <td>If broker configs are altered, then the broker will check topic level privileges.</td>
```
same as above

omkreddy commented on docs/security.html:
```diff
@@ -1258,6 +1258,445 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <td></td>
+        </tr>
+        <tr>
+            <td>METADATA (3)</td>
```
This is same as above entry?

viktorsomogyi replied:
There is a typo, there supposed to be a Topic resource instead of a Cluster. So if topic auto-creation is enabled, then the handleTopicMetadataRequest first checks for privileges on the Cluster resource, then if there is none found checks on the Topic resource. I'll add a description explaining this as Mickael suggested and fix this.

omkreddy commented on docs/security.html:
```diff
@@ -1258,6 +1258,445 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <td>LIST_GROUPS (16)</td>
+            <td>Describe</td>
+            <td>Group</td>
+            <td>If none of the groups are authorized, then just an empty response will be sent back instead
```
Can we mention that this is applicable from 2.1 release?

omkreddy commented on docs/security.html:
```diff
@@ -1258,6 +1258,445 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <td>CREATE_TOPICS (19)</td>
+            <td>Create</td>
+            <td>Topic</td>
+            <td></td>
```
Can we mention that this is applicable from 2.0 release?

omkreddy reviewed (commented):
@viktorsomogyi  Thanks for the Patch. Left few minor comments.

omkreddy commented on docs/security.html:
```diff
@@ -1258,6 +1258,445 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <td>OFFSET_FOR_LEADER_EPOCH (23)</td>
+            <td>Describe</td>
+            <td>Topic</td>
+            <td></td>
```
Can we mention that this is required from 2.1 release?

viktorsomogyi replied:
For now I'll add these version notes but what do you think about adding a version column at the end in a separate pull request? (I wouldn't add this column in this PR just to limit the scope now and get out something that is already useful.)

mimaison commented on docs/security.html:
```diff
@@ -1258,6 +1258,445 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <td>METADATA (3)</td>
+            <td>Create</td>
+            <td>Cluster</td>
+            <td></td>
```
The `Create` permissions is only needed to the topic does not exist, auto-creation is allowed, and the metadata has the `allow_auto_topic_creation` flag set

mimaison commented on docs/security.html:
```diff
@@ -1258,6 +1258,445 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <td>Write</td>
+            <td>Topic</td>
+            <td></td>
+        </tr>
```
Instead of having 3 times `PRODUCE (0)`, can we can we add the use case each entry covers? like in https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Authorizations

For example, "normal" produce requires the 3rd entry only, while Idempotent produce requires both the 2nd and 3rd.

viktorsomogyi replied:
Yeah I was planning to do that, so I guess I'll just do it know :)

mimaison commented on docs/security.html:
```diff
@@ -1258,6 +1258,445 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <td>Describe</td>
+            <td>TransactionalId</td>
+            <td></td>
+        </tr>
```
Can we mention that this applies only to Idempotent/transactional producers?

mimaison reviewed (changes requested):
Thanks for adding this in the official documentation. Currently it's not easy to implement a custom authorizer and having the ACL breakdown will definitively help. However I think the table can be a bit confusing in its current format, see comments

viktorsomogyi committed:
Adding some extra notes for certain privileges

viktorsomogyi commented:
@mimaison, @omkreddy I think I've addressed the issues you raised. Please let me know if I missed something.

omkreddy commented on docs/security.html:
```diff
@@ -1258,6 +1258,449 @@ <h4><a id="security_authz_examples" href="#security_authz_examples">Examples</a>
+            <td>This authorizes topic auto-creation on Topic resources if enabled.</td>
+        </tr>
+        <tr>
+            <td>METADATA (3)</td>
```
this looks like same as above entry?

viktorsomogyi replied:
Fixed it, thanks. Went through my list searching for other duplicates but I think this is the last one.

viktorsomogyi committed:
Remove duplicate entry and add some more description to other entries

mimaison reviewed (approved):
LGTM! Thanks

omkreddy reviewed (approved):
@viktorsomogyi Thanks for the PR. LGTM

omkreddy commented:
@ijuma @hachikuji  Can you take a quick look?

omkreddy committed:
fix minor spelling corrections

omkreddy commented:
Merging to trunk. Any further comments can be addressed in subsequent PRs.

ijuma commented:
@omkreddy was this merged without any committer review?

omkreddy commented:
@ijuma I reviewed and tested this doc PR raised by @viktorsomogyi . I pushed a trivial change to correct spelling mistakes.  I went ahead and merged this doc PR.
Please correct me, If I deviated from the merge process.

ijuma commented:
@omkreddy Not at all, you followed the process. I got a little confused because I thought you were asking for a review via https://github.com/apache/kafka/pull/5906#issuecomment-442411689, but now I realise that you were asking for an additional review. Thanks for the clarification.
